spinalcordtoolbox/reports/qc2.py

```diff
@@ -0,0 +1,352 @@
+        fig.set_size_inches(*size_fig, forward=True)
+        FigureCanvas(fig)
+        ax = fig.add_axes((0, 0, 1, 1))
+        ax.imshow(img, cmap='gray', interpolation='none', aspect=1.0)
+        add_orientation_labels(ax)
+        ax.get_xaxis().set_visible(False)
+        ax.get_yaxis().set_visible(False)
+        img_path = str(path_background_img)
+        logger.debug('Save image %s', img_path)
+        fig.savefig(img_path, format='png', transparent=True, dpi=300)
+
+        # Generate the second QC report image
+        img = equalize_histogram(mosaic(img_output, centers))
+        fig = Figure()
+        fig.set_size_inches(*size_fig, forward=True)
+        FigureCanvas(fig)
+        ax = fig.add_axes((0, 0, 1, 1), label='0')
+        ax.imshow(img, cmap='gray', interpolation='none', aspect=1.0)
+        add_orientation_labels(ax)
+        ax.get_xaxis().set_visible(False)
+        ax.get_yaxis().set_visible(False)
+        img_path = str(path_overlay_img)
+        logger.debug('Save image %s', img_path)
+        fig.savefig(img_path, format='png', transparent=True, dpi=300)
+
+
+def inf_nan_fill(A: np.ndarray):
+    """
+    Interpolate inf and NaN values with neighboring values in a 1D array, in-place.
+    If only inf and NaNs, fills the array with zeros.
+    """
+    valid = np.isfinite(A)
+    invalid = ~valid
+    if np.all(invalid):
+        A.fill(0)
+    elif np.any(invalid):
+        A[invalid] = np.interp(
+            np.nonzero(invalid)[0],
+            np.nonzero(valid)[0],
+            A[valid])
+
+
+def mosaic(img: Image, centers: np.ndarray, radius: tuple[int, int] = (15, 15)):
+    """
+    Arrange the slices of `img` into a grid of images.
+
+    Each slice is centered at the approximate coordinates given in `centers`,
+    and cropped to `radius` pixels in each direction (horizontal, vertical).
+
+    If `img` has N slices, then `centers` should have shape (N, 2).
+    """
+    # Fit as many slices as possible in each row of 600 pixels
+    num_col = math.floor(600 / (2*radius[0]))
+
+    # Center and crop each axial slice
+    cropped = []
+    for center, slice in zip(centers.astype(int), img.data):
+        # Add a margin before cropping, in case the center is too close to the edge
+        cropped.append(np.pad(slice, radius)[
+            center[0]:center[0] + 2*radius[0],
+            center[1]:center[1] + 2*radius[1],
+        ])
+
+    # Pad the list with empty arrays, to get complete rows of num_col
+    empty = np.zeros((2*radius[0], 2*radius[1]))
+    cropped.extend([empty] * (-len(cropped) % num_col))
+
+    # Arrange the images into a grid
+    return np.block([cropped[i:i+num_col] for i in range(0, len(cropped), num_col)])
+
+
+def add_orientation_labels(ax: Axes):
+    """
+    Add orientation labels (A, P, L, R) to a figure, yellow with a black outline.
+    """
+    for x, y, letter in [
+        (12, 6, 'A'),
+        (12, 28, 'P'),
+        (0, 18, 'L'),
+        (24, 18, 'R'),
```

Comment:
Clearly this should be customizable. And in this proof-of-concept, the caller of this function would actually have enough context to know which labels to put where (unlike in the existing code structure).

Reply:
Specifically, this ties into an earlier issue we encountered:

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4122

Reply:
Exactly. Thanks for the cross-reference.